### PR TITLE
#2343 Add possibility to use a dynamic model on bus and avoid network…

### PR DESCRIPTION
**Pull request by gautierbureau**
… creation if dyd covers all iidm equipments.

**Checklist before requesting a review**

*use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes*

- [x] unit tests and non regression tests were added (new model, new feature and bug fix)
- [ ] main documentation was updated (update of input/output file, 3rd party, model, repository organization, solver)
- [ ] example documentations were updated (new example in examples folder)
- [ ] if this PR modifies the parameters or inputs/outputs of a model/solver: the corresponding xsl was added in util/xsl and platform DB were updated
- [x] if this PR modifies a dictionary: the corresponding french dictionary was updated
- [ ] if this commit targets a release branch: the corresponding milestone was added in the ticket and in this PR

### CODE REVIEW

File: dynawo/sources/Modeler/DataInterface/DYNDataInterface.h [hunk 1]
```diff
@@ -205,6 +205,12 @@ class DataInterface {
    * @returns cloned data interface
    */
   virtual boost::shared_ptr<DataInterface> clone() const = 0;
+
+  /**
+  * @brief test if some network components does not have a dynamic model
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/DYNDataInterface.h:210]: Please explain in which case the network is instantiated or not

File: dynawo/sources/Modeler/DataInterface/DYNDataInterfaceIIDMCommon.cpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
+//
+// Copyright (c) 2022, RTE (http://www.rte-france.com)
+// See AUTHORS.txt
+// All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at http://mozilla.org/MPL/2.0/.
+// SPDX-License-Identifier: MPL-2.0
+//
+// This file is part of Dynawo, an hybrid C++/Modelica open source time domain
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/DYNDataInterfaceIIDMCommon.cpp:10]: Not the latest description of dynawo
`// This file is part of Dynawo, an hybrid C++/Modelica open source suite of simulation tools

// for power systems.`

File: dynawo/sources/Modeler/DataInterface/DYNDataInterfaceIIDMCommon.cpp [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+using boost::shared_ptr;
+using std::vector;
+
+namespace DYN {
+
+bool
+DataInterfaceIIDM::instantiateNetwork() const {
+  shared_ptr<NetworkInterface> network = getNetwork();
+  const vector<shared_ptr<VoltageLevelInterface> >& voltageLevels = network->getVoltageLevels();
+  vector<shared_ptr<VoltageLevelInterface> >::const_iterator iVL;
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/DYNDataInterfaceIIDMCommon.cpp:39]: Please insert the iterator declaration Inside the for loop (same everywhere)

File: dynawo/sources/Modeler/DataInterface/IIDM/CMakeLists.txt [hunk 4]
```diff
@@ -38,6 +38,7 @@ set(DATA_INTERFACE_IIDM_SOURCES
     DYNLccConverterInterfaceIIDM.cpp
     DYNHvdcLineInterfaceIIDM.cpp
     DYNServiceManagerInterfaceIIDM.cpp
+    ../DYNDataInterfaceIIDMCommon.cpp
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/IIDM/CMakeLists.txt:41]: Please avoid relative paths in cmakes

File: nrt/data/SMIB/SMIB_BasicTestCases/SMIB_1_StepPm_IIDM/SMIB.dyd [hunk 5]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright (c) 2015-2019, RTE (http://www.rte-france.com)
```
Comment (rosiereflo) [inline nrt/data/SMIB/SMIB_BasicTestCases/SMIB_1_StepPm_IIDM/SMIB.dyd:3]: 2022 + description of dynawo (same everywhere)

File: nrt/data/SMIB/SMIB_BasicTestCases/SMIB_1_StepPm_IIDM/SMIB.par [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    <par type="DOUBLE" name="URef_ValueIn" value="0"/>
+    <par type="DOUBLE" name="Pm_ValueIn" value="0"/>
+    <par type="DOUBLE" name="Pm_Height" value="0.02"/>
+    <par type="DOUBLE" name="Pm_tStep" value="1"/>
+  </set>
+  <!-- OmegaRef -->
+  <set id="6">
+    <par type="DOUBLE" name="setPoint_Value0" value="1"/>
+  </set>
+  <set id="7">
```
Comment (rosiereflo) [inline nrt/data/SMIB/SMIB_BasicTestCases/SMIB_1_StepPm_IIDM/SMIB.par:99]: Why do we need this if the network is not used?

**Review by rosiereflo — CHANGES_REQUESTED** [review]
- French dictionary update is missing
- I think we should add this feature in the documentation (in the section of the NETWORK model?)

File: dynawo/sources/Modeler/DataInterface/DYNDataInterface.h (hunk 1 above)
Comment (gautierbureau) [reply dynawo/sources/Modeler/DataInterface/DYNDataInterface.h:210]: ok

File: dynawo/sources/Modeler/DataInterface/DYNDataInterfaceIIDMCommon.cpp (hunk 2 above)
Comment (gautierbureau) [reply dynawo/sources/Modeler/DataInterface/DYNDataInterfaceIIDMCommon.cpp:10]: done

File: nrt/data/SMIB/SMIB_BasicTestCases/SMIB_1_StepPm_IIDM/SMIB.dyd (hunk 5 above)
Comment (gautierbureau) [reply nrt/data/SMIB/SMIB_BasicTestCases/SMIB_1_StepPm_IIDM/SMIB.dyd:3]: done

File: nrt/data/SMIB/SMIB_BasicTestCases/SMIB_1_StepPm_IIDM/SMIB.par (hunk 6 above)
Comment (gautierbureau) [reply nrt/data/SMIB/SMIB_BasicTestCases/SMIB_1_StepPm_IIDM/SMIB.par:99]: good point !

File: dynawo/sources/Modeler/DataInterface/DYNDataInterfaceIIDMCommon.cpp (hunk 3 above)
Comment (gautierbureau) [reply dynawo/sources/Modeler/DataInterface/DYNDataInterfaceIIDMCommon.cpp:39]: done

File: dynawo/sources/Modeler/DataInterface/IIDM/CMakeLists.txt (hunk 4 above)
Comment (gautierbureau) [reply dynawo/sources/Modeler/DataInterface/IIDM/CMakeLists.txt:41]: The issue is that the implem is valid for both IIDM library and then I try to make it common for both and not store it in the subfolders

File: dynawo/sources/Modeler/DataInterface/DYNDataInterfaceIIDMCommon.cpp (hunk 3 above)
Comment (gautierbureau) [reply dynawo/sources/Modeler/DataInterface/DYNDataInterfaceIIDMCommon.cpp]: I refactored the code entirely

File: nrt/data/SMIB/SMIB_BasicTestCases/SMIB_1_StepPm_IIDM/SMIB.jobs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    SPDX-License-Identifier: MPL-2.0
+
+    This file is part of Dynawo, an hybrid C++/Modelica open source suite of simulation tools
+    for power systems.
+-->
+<dyn:jobs xmlns:dyn="http://www.rte-france.com/dynawo">
+  <dyn:job name="SMIB_StepPm_IIDM">
+      <dyn:solver lib="dynawo_SolverIDA" parFile="solvers.par" parId="5"/>
+      <dyn:modeler compileDir="outputs/compilation">
+        <dyn:network iidmFile="SMIB.iidm" parFile="SMIB.par" parId="7"/>
```
Comment (rosiereflo) [inline nrt/data/SMIB/SMIB_BasicTestCases/SMIB_1_StepPm_IIDM/SMIB.jobs:18]: Can we remove the parameters here?

File: dynawo/sources/Modeler/DataInterface/IIDM/DYNDataInterfaceIIDM.cpp [hunk 8]
```diff
@@ -1226,7 +1226,7 @@ DataInterfaceIIDM::findConnectedComponents() {
   shared_ptr<vector<shared_ptr<ComponentInterface> > > connectedComponents = boost::make_shared<vector<shared_ptr<ComponentInterface> > >();
   for (map<string, shared_ptr<ComponentInterface> >::iterator iter = components_.begin(), iterEnd = components_.end();
        iter != iterEnd; ++iter) {
-    if ((iter->second)->isPartiallyConnected())
+    if ((iter->second)->isConnected())
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/IIDM/DYNDataInterfaceIIDM.cpp:1229]: Need a rebase?

File: dynawo/sources/Modeler/DataInterface/IIDM/CMakeLists.txt (hunk 4 above)
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/IIDM/CMakeLists.txt:41]: Then maybe it should be in the dynawo_DataInterface library?

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: nrt/data/SMIB/SMIB_BasicTestCases/SMIB_1_StepPm_IIDM/SMIB.jobs (hunk 7 above)
Comment (gautierbureau) [reply nrt/data/SMIB/SMIB_BasicTestCases/SMIB_1_StepPm_IIDM/SMIB.jobs:18]: The parFile and parId are required in the xsd. Maybe we can open a new issue to make them optional ?

File: dynawo/sources/Modeler/DataInterface/DYNDataInterface.h [hunk 9]
```diff
@@ -208,6 +208,9 @@ class DataInterface {
 
   /**
   * @brief test if some network components does not have a dynamic model
+  *
+  * A network model will be instantiated if only one of all the static components does not have a dynamic model.
```
[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/sources/Modeler/DataInterface/DYNDataInterface.h:212]: done

File: dynawo/sources/Modeler/DataInterface/IIDM/DYNDataInterfaceIIDM.cpp (hunk 8 above)
Comment (gautierbureau) [reply dynawo/sources/Modeler/DataInterface/IIDM/DYNDataInterfaceIIDM.cpp:1229]: done

File: nrt/data/SMIB/SMIB_BasicTestCases/SMIB_1_StepPm_IIDM/SMIB.jobs (hunk 7 above)
Comment (rosiereflo) [reply nrt/data/SMIB/SMIB_BasicTestCases/SMIB_1_StepPm_IIDM/SMIB.jobs:18]: ok

File: dynawo/sources/Modeler/DataInterface/IIDM/CMakeLists.txt (hunk 4 above)
Comment (gautierbureau) [reply dynawo/sources/Modeler/DataInterface/IIDM/CMakeLists.txt]: done

File: dynawo/sources/Modeler/DataInterface/IIDM/DYNDataInterfaceIIDM.cpp [hunk 10]
```diff
@@ -1226,7 +1226,7 @@ DataInterfaceIIDM::findConnectedComponents() {
   shared_ptr<vector<shared_ptr<ComponentInterface> > > connectedComponents = boost::make_shared<vector<shared_ptr<ComponentInterface> > >();
   for (map<string, shared_ptr<ComponentInterface> >::iterator iter = components_.begin(), iterEnd = components_.end();
        iter != iterEnd; ++iter) {
-    if ((iter->second)->isConnected())
+    if ((iter->second)->isPartiallyConnected())
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/DataInterface/IIDM/DYNDataInterfaceIIDM.cpp:1229]: Unwanted change I guess

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Modeler/DataInterface/IIDM/DYNDataInterfaceIIDM.cpp (hunk 10 above)
Comment (gautierbureau) [reply dynawo/sources/Modeler/DataInterface/IIDM/DYNDataInterfaceIIDM.cpp]: it's ok now

**Review by rosiereflo — APPROVED** [review]

**Pull request merged**